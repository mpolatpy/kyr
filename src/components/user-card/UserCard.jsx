import React from 'react';
import {Link} from 'react-router-dom';
import { connect } from 'react-redux';
import { deleteUser } from '../../redux/users/usersActions';
import Box from '@mui/material/Box';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import EditIcon from '@mui/icons-material/Edit';
import DeleteIcon from '@mui/icons-material/Delete';
import Tooltip from '@mui/material/Tooltip';

const UserCard = ({ user, deleteUser }) => {

    if(!user) return null;

    return (
        <Box sx={{ width: '80vw', maxWidth: 600, margin: '10px' }}>
            <Card variant="outlined">
                <CardContent>
                    <Typography variant="h6" gutterBottom>
                        {user.name}
                    </Typography>
                    <Typography>
                        Email: {user.email.toLowerCase()}
                    </Typography>
                </CardContent>
                <CardActions>
                    <Tooltip title="Edit User">
                        <IconButton 
                            component={Link}  
                            to={`/users/${user.id}`}
                            aria-label="edit-details"
                        >
                            <EditIcon />
                        </IconButton>
                    </Tooltip>
                    <Tooltip title="Delete User">
                        <IconButton
                            onClick={() => deleteUser(user.id)} 
                            aria-label="delete-user"
                        >
                            <DeleteIcon />
                        </IconButton>
                    </Tooltip>
                </CardActions>
            </Card>
        </Box>
    );
}

const mapDispatchToProps = dispatch => ({
    deleteUser: (id) => dispatch(deleteUser(id))
})

export default connect(null, mapDispatchToProps)(UserCard);
