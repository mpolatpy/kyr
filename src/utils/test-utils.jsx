import React from 'react'
import { render } from '@testing-library/react'
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux'
import rootReducer from '../redux/rootReducer';

const thunk =
  ({ dispatch, getState }) =>
  next =>
  action => {
    if (typeof action === 'function') {
      return action(dispatch, getState)
    }

    return next(action)
  }

  

export const rdxRender = (
    ui,
    { initialState, ...renderOptions } = {}
  ) => {
    const store = createStore(rootReducer, initialState, applyMiddleware(thunk));
    const Wrapper = ({ children }) => (
      <Provider store={store}>{children}</Provider>
    );
  
    return render(ui, { wrapper: Wrapper, ...renderOptions });
  };