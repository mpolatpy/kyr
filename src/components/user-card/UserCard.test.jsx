import React from 'react';
import '@testing-library/jest-dom'
import { screen, fireEvent } from '@testing-library/react';
import { rdxRender } from '../../utils/test-utils';
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import UserCard from "./UserCard";

describe('UserCard Component Tests', () => {
    let wrapper, mockDispatch, history;

    beforeEach(() => {
        history = createMemoryHistory();
        mockDispatch = jest.fn();

        const mockUser = {
            name: 'Michael Jordan',
            email: 'mjordan@kyrrus.com',
            id: '125'
        }

        const mockProps = {
            user: mockUser,
            dispatch: mockDispatch
        }

        wrapper = rdxRender(
            <Router location={history.location} navigator={history}>
                <UserCard {...mockProps} />
            </Router>,
            {});
    })

    test('it should match snaphot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Edit Icon should be in the document and navigate to user edit page', () => {
        const editIcon = screen.getByRole('link', { name: 'edit-details' });
        expect(editIcon).toBeInTheDocument();

        fireEvent.click(editIcon)
        expect(history.location.pathname).toBe('/users/125')
    });

    test('Delete Icon should be in the document', () => {
        const deleteIcon = screen.getByRole('button', { name: 'delete-user' });
        expect(deleteIcon).toBeInTheDocument();
    });

})