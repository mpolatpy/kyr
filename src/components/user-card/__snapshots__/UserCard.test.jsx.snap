// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard Component Tests it should match snaphot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-16f3ot7"
      >
        <div
          class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
            >
              Michael Jordan
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Email: 
              mjordan@kyrrus.com
            </p>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <a
              aria-label="edit-details"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              href="/users/125"
              tabindex="0"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </a>
            <button
              aria-label="delete-user"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-16f3ot7"
    >
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
          >
            Michael Jordan
          </h6>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Email: 
            mjordan@kyrrus.com
          </p>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
        >
          <a
            aria-label="edit-details"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element="true"
            href="/users/125"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
          <button
            aria-label="delete-user"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element="true"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="DeleteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
